File: 98. Validate Binary Search Tree.java
// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    TreeNode prev = null;
    public boolean isValidBST(TreeNode root) {
        return inorder(root);
    }
    
    public boolean inorder(TreeNode root){
        //base
        if(root == null ) return true;
        //logic
        if(!inorder(root.left)) return false;
        if(prev!=null && prev.val>=root.val) return false;
        
        prev = root;
        return inorder(root.right);
        
    }
    //Approach1
//     public boolean isValidBST(TreeNode root) {
//         return checkBST(root,Long.MIN_VALUE,Long.MAX_VALUE);
//     }
//     public boolean checkBST(TreeNode root, long min, long max){
//         if(root == null) return true;
        
//         if( min>=root.val ) return false;
//         if( max<=root.val) return false;
        
//         return checkBST(root.left,min,root.val) && checkBST(root.right,root.val,max);
//     }
}

File: 98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

/*
Time Complexity: O(N)
Space Complexity: O(1)

Intution: An inorder traversal of a BST will result in a strictly-increasing array (if the values are unique)
So maintain the previous encountered value and compare with the current node's value at each traverse 
*/
class Solution {
public:
    bool set;
    int prev;

    bool call(TreeNode* root){
      if(root==nullptr)  return true;

      bool left = call(root->left);

      //check if strictly increasing
      if(set && root->val <= prev)  return false; //set = flag to indicate if a value has been encountered yet
      
      prev = root->val;
      set = true;

      bool right = call(root->right);

      return left && right;
    }

    bool isValidBST(TreeNode* root) {
        set = false;
        return call(root);
    }
};
